Adds PyTorch model management

joshdevins opened the pull request:
This adds PyTorch model management to eland. This is a clean branch with changes made from our last discussions. The main thing to continue working on in this PR is how we expose the script `upload_hub_model.py` since it should be "installed" and easily run with a simple command. Happy to discuss other items inline as well.

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+
+    def start(self, timeout: str = DEFAULT_TIMEOUT) -> bool:
+        return self._client.transport.perform_request(
+            method="POST",
```
Let's move the `method` and `url` options to be positional:

```python
self._client.transport.perform_request("POST", f"/_ml/...", params={...})
```

Also the return type here is `bool` but this function either raises an exception or returns a JSON body. Perhaps we should switch to `-> None` and drop the `return`?

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+        try:
+            return self._client.transport.perform_request(
+                method="POST",
+                url=f"/_ml/trained_models/{self.model_id}/deployment/_stop",
```
Same as above. Also let's remove the `ignore: 404` as we're catching the exception.

Can we remove the return and return type on this method as well?

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+        except elasticsearch.NotFoundError:
+            pass
+
+    def delete(self) -> Dict[str, Any]:
```
Same comments as above

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+            pass
+
+    @staticmethod
+    def list(
```
Since this isn't doing a ton beyond what the client is already doing, thoughts on removing this method?

joshdevins replied:
I wanted to have start, stop, delete all in the same place since right now the client doesn't have start and stop on trained model. So having them all here is just consistent until they are all available in the client.

joshdevins replied:
I can remove `list` since I only really use it in a "cleanup" script for QA.

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+
+        return True
+
+    def upload(
```
Let's switch to a `@classmethod` called `import_model` to match with the existing `eland.ml.MLModel` class?

https://github.com/elastic/eland/blob/main/eland/ml/ml_model.py#L239

joshdevins replied:
I tried this and it doesn't quite feel right. I prefer creating a `PyTorchModel` first so a user can choose to check for existing models, stop it, etc. This obviates the need for `es_exists_model` parameter like in `MLModel`.

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+
+    def upload(
+        self,
+        model_path: str,
```
I'm not sure if it makes sense to do this but a lot of times when users are presented with an API that only takes paths to files they immediately as for "can I do it with data in memory"? Does that even make sense here, if so I have an addition review comment to ask if we can support "path or loaded JSON"

joshdevins replied:
Yes, definitely. I might need some help making this look right. In the case of model upload, I need a `BinaryIO`. For the configs, it can be easily a path to load or a `Dict` already loaded.

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+        file_stats = os.stat(model_path)
+        total_parts = math.ceil(file_stats.st_size / chunk_size)
+
+        def model_file_chunk_generator():
```
Let's add type hints to inner functions too. ` -> Iterable[str]`

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+                    yield base64.b64encode(data).decode()
+
+        for i, data in tqdm(
+            enumerate(model_file_chunk_generator(), start=0), total=total_parts
```
nit: start=0 is the default so not needed

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+                        break
+                    yield base64.b64encode(data).decode()
+
+        for i, data in tqdm(
```
Should we do `tqdm.auto...` here?

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+    def _upload_model(
+        self, model_path: str, chunk_size: int = DEFAULT_CHUNK_SIZE
+    ) -> bool:
+        file_stats = os.stat(model_path)
```
nit: Since `file_stats.st_size` appears to be the only thing used perhaps we can save that value instead of the whole stats object?

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+        self.model_id = model_id
+
+    @staticmethod
+    def _load_json(path: str) -> Dict[str, Any]:
```
Can we inline `with open(path) as f: config=json.load(f)`?

sethmlarson reviewed (changes requested):
Collection of comments regarding only the `PytorchModel` implementation

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+#  Licensed to Elasticsearch B.V. under one or more contributor
```
Also can we make this module private like `_pytorch_model.py` so users know not to import things from here?

sethmlarson commented on eland/ml/pytorch/pytorch_model.py:
```diff
@@ -0,0 +1,154 @@
+    from elasticsearch import Elasticsearch
+
+DEFAULT_CHUNK_SIZE = 4 * 1024 * 1024  # 4MB
+QUEUE_SIZE = 4
```
Some of these variables aren't used?

joshdevins commented:
@sethmlarson Have a peek again, I've addressed most of the comments (and our Slack discussion to cleanup the scripts). There are a one or two unresolved issues that I need to chat about/get help with still.

sethmlarson reviewed (changes requested):
Few small comments then I think we can merge this and iterate on the CLI location+config in another PR.

sethmlarson commented:
Let's get it merged then, I can review PRs but I won't have any time to do the development myself unfortunately as I'm slammed with 8.0 work.
